src/main/telemetry/ibus_shared.c

```diff
@@ -57,120 +79,460 @@ typedef enum {
+    for (uint8_t i = 0; i < sizeof(GPS_IDS); i++) {
+        if (sensorType == GPS_IDS[i]) {
+            result = true;
+            break;
         }
-        return sendIbusMeasurement(address, value);
+    }
+    if (!result) return result;
+
+    uint16_t gpsFixType = 0;
+    uint16_t sats = 0;
+    if (sensors(SENSOR_GPS)) {
+        gpsFixType = !STATE(GPS_FIX) ? 1 : (gpsSol.numSat < 5 ? 2 : 3);
+        sats = gpsSol.numSat;
+        if (STATE(GPS_FIX) || sensorType == IBUS_SENSOR_TYPE_GPS_STATUS) {
+            result = true;
+            switch (sensorType) {
+            case IBUS_SENSOR_TYPE_SPE:
+                value->uint16 = gpsSol.groundSpeed * 36 / 100;
+                break;
+            case IBUS_SENSOR_TYPE_GPS_LAT:
+                value->int32 = gpsSol.llh.lat;
+                break;
+            case IBUS_SENSOR_TYPE_GPS_LON:
+                value->int32 = gpsSol.llh.lon;
+                break;
+            case IBUS_SENSOR_TYPE_GPS_ALT:
+                value->int32 = (int32_t)gpsSol.llh.alt;
+                break;
+            case IBUS_SENSOR_TYPE_GROUND_SPEED:
+                value->uint16 = gpsSol.groundSpeed;
+                break;
+            case IBUS_SENSOR_TYPE_ODO1:
+            case IBUS_SENSOR_TYPE_ODO2:
+            case IBUS_SENSOR_TYPE_GPS_DIST:
+                value->uint16 = GPS_distanceToHome;
+                break;
+            case IBUS_SENSOR_TYPE_COG:
+                value->uint16 = gpsSol.groundCourse * 100;
+                break;
+            case IBUS_SENSOR_TYPE_GPS_STATUS:
+                value->byte[0] = gpsFixType;
+                value->byte[1] = sats;
+                break;
+            }
+        }
+    }
+    return result;
+}
+#endif //defined(GPS)
+
+static void setValue(uint8_t* bufferPtr, uint8_t sensorType, uint8_t length){
 
-    case IBUS_SENSOR_TYPE_TEMPERATURE:
-        value = gyroGetTemperature() * 10;
-        return sendIbusMeasurement(address, value + IBUS_TEMPERATURE_OFFSET);
+    ibus_telemetry_t value;
 
-    case IBUS_SENSOR_TYPE_RPM:
-        return sendIbusMeasurement(address, (uint16_t) rcCommand[THROTTLE]);
+#if defined(TELEMETRY_IBUS_EXTENDED)
+    const uint8_t* structure = 0;
+    uint8_t itemCount;
+    if (sensorType == IBUS_SENSOR_TYPE_GPS_FULL) {
+        structure = FULL_GPS_IDS;
+        itemCount = sizeof(FULL_GPS_IDS);
+    }
+    if (sensorType == IBUS_SENSOR_TYPE_VOLT_FULL) {
+        structure = FULL_VOLT_IDS;
+        itemCount = sizeof(FULL_VOLT_IDS);
+    }
+    if (sensorType == IBUS_SENSOR_TYPE_ACC_FULL) {
+        structure = FULL_ACC_IDS;
+        itemCount = sizeof(FULL_ACC_IDS);
+    }
+    if (structure != 0) {
+        setCombinedFrame(bufferPtr, structure, sizeof(itemCount));
+        return;
     }
-    return 0;
+#endif //defined(TELEMETRY_IBUS_EXTENDED)
+
+#if defined(GPS)
+    if (setGPS(sensorType, &value)) {
+        return;
+    }
+#endif //defined(TELEMETRY_IBUS_EXTENDED)
+
+    for (uint8_t i = 0; i < length; i++) {
```

Comment:
Use `unsigned` here, it's more efficient, because it avoids overflow checking.

Reply:
In my code all loops are using unsigned 8 bits integers - should be fine. The variable length is also of same type.